04_Scheduling/02_Scheduler.md

```diff
@@ -154,14 +154,64 @@ void schedule() {
+    while(tasks_list[current_executing_task_idx] == NULL) {        
+        current_executing_task_idx = (current_executing_task_idx + 1) % MAX_TASKS;
+    }
+}
+
+```
+
+that's it, this is how how we save the status of the current executing task. 
+
+#### Restoring context
+
+After the previous task status has been saved and the next task has been selected we are now ready to make the switch. How to do it is pretty straightforward, but let's have a quick recap on how we are handling interrupts: 
+
+1. The first thing is that whenever an interrupt is fired an asm code snippet is called saves the current cpu registers status, on the stack, 
+2. Then once the status is saved, it place the current stack pointer (`rsp`) on the `rdi` register (it is the first c function parameter for x86_64 architecture) and call the interrupt handler. This parameter is our `context` variable (that is passed also to the scheduler)
+3. Then our interrupt handler function (unless we are writing an os entirely in asm, this usually is on our language of choice, in our case C)  does whatever it needs to serve the interrupt  and  the context variable. Returning a variable, according to the ABI calling convention is corresponding to placing the returning value to the `rax` register
+4. Now we are back in the asm code where `rax` value is placed on the `rsp` register, and then the context restored...
+5. and everyone lived happily after (until the next interrupt will happens...) 
+
+So what is happening is that we pass the context as a parameter to the interrupt handler routine, and when we leave it we return it and use the returned value to restore the cpu status just before resuming the normal execution. 
+
+And the context being passed is the stack pointer, now the whole idea behind the interrupt handling routine is that we save the context on the stack before starting the interrupt handler, and restore it once the interrupt is done. So what happens if instead of restoring the context from the same stack pointer, we restore it from a different one? 
+
+What will happens is that it resumes the execution of whatever was stored in that other context (if it was a valid one), hence we have a *task switch*. 
+
+And to implement this change we just need to make two tiny adjustments to our scheduler function: change its return type from `void` to `cpu_status_t*` and return the newly loaded context to the caller: 
+
+```c
+cpu_status_t* schedule(cpu_status_t* context) {
+    tasks_list[current_executing_task_idx].context = context
+    current_executing_task_idx = (current_executing_task_idx + 1) % MAX_TASKS;
+    while(tasks_list[current_executing_task_idx] == NULL) {        
+        current_executing_task_idx = (current_executing_task_idx + 1) % MAX_TASKS;
+    }
+    return tasks_list[current_executing_task_idx].context;
+}
+```
+
+And now our scheduler is capable of switching tasks. /* should i mention that this should work only if the rsp is changed with the new stack? */
```

Comment:
I would say yes, but you did mention above that this expects the reader has implemented an interrupt handler similar to one we implemeneted in the interrupts chapter. I think that's enough of a hint, but if you want to be sure a sentence on it here wouldn't hurt :)